src/ssl/support.h

```diff
@@ -86,7 +86,7 @@ bool InitServerContext(Security::ContextPointer &, AnyP::PortCfg &);
 bool InitClientContext(Security::ContextPointer &, Security::PeerOptions &, long flags);
 
 /// set the certificate verify callback for a context
-void SetupVerifyCallback(Security::ContextPointer &);
+void SetupPeerVerification(Security::ContextPointer &, bool, bool);
```

Comment:
You have decided to use two boolean parameters instead of one flags parameter. Correct usage (and reading/review) of these two boolean parameters requires familiarization with the SetupPeerVerification() API details (well, in your particular implementation, it actually requires learning about that API _implementation_, but that problem is trivial to fix if we stay with this design). It is easy to accidentally swap them and/or copy the wrong old parameter when writing new code because many developers and reviewers are unlikely to notice such mistakes. Such mistakes have happened in the past with similar APIs.

Is it very difficult to use a single flags parameter instead? If not, please do that.

Reply:
I'm trying to limit the API to limit the possibility of bad values needing some sort of error handling which would require a more significant change. How about an enum so at least there would be warnings if misused? If using flags might as well just use the SSL flags.

Reply:
I hope there is some misunderstanding here. I reordered your statements to clarify my suggestion first.

> If using flags might as well just use the SSL flags.

I am not sure what you mean by "SSL flags" but my suggestion was (and still is) to pass `parsedFlags`. As is. Instead of two post-flags-analysis booleans.  SetupPeerVerification() itself will perform all the necessary analysis to form the right OpenSSL mode parameter. Sorry if that was not clear. AFAICT, the code will look much simpler after that change and addressing other problems mentioned in my review may become easier as well.

> I'm trying to limit the API to limit the possibility of bad values needing some sort of error handling

The flags-based API I have proposed does not imply any unnecessary error handling AFAICT.

> How about an enum so at least there would be warnings if misused?

I would not recommend that for two reasons: 

1. We may end up with code that converts Squid configuration flags into internal enums _and_ code that converts those internal enums to OpenSSL configuration flags. Hopefully, we do not need such back-and-forth conversions!
1. Enum itself is a poor medium for representing nested/dependent flags (as illustrated by OpenSSL enum-like API effectively causing a few other PR problems marked in my review).

Reply:
I mean the literal SSL_VERIFY mode flags. I was asked to abstract the call to SSL_CTX_set_verify.  I'm confused as to what that means now. There are only 3 legitimate values... SSL_VERIFY_NONE, SSL_VERIFY_PEER, and SSL_VERIFY_PEER|SSL_VERIFY_PEER_IF_NO_PEER_CERT. The boolean API style didn't protect against just SSL_VERIFY_PEER_IF_NO_PEER_CERT because it is, while a bug, apparently a legit client mode value as mentioned in the man page.

The SSL_VERIFY_* flags are a mix of actual flags and enums in practice. NONE = 0.

I see only 4 ways to pass in the necessary information. 1 function with 3 different argument styles (booleans, enums, or bit flags) and using 3 different functions. Bitflags are what existed prior and I was asked to remove the direct SSL_CTX_set_verify calls. Booleans make it harder to tell what's going on. Separate functions is what I had the first time and were rejected. And enums are also said to be no good.

Enums or 3 different functions are the safest means to limit the selection. If the suggestion is simply to pass around SSL_VERIFY mode flags raw I'm not understanding the request for the abstraction since there really isn't any. At most there would be the ability for uniform error handling which would be a different kind of change and I'd rather not mix restructuring with improvements. If we create new bit flags to provide an abstraction then we'll need more conversion code than enums would.

Please just tell me precisely the function signature you want.

Reply:
Additionally, `parsedFlags` isn't available in support.cc. So I'd need a different interface or recreate the necessary flags to call the function.

Reply:
How should I handle the `clientCaStack` conditional affecting the verification mode? I either have to have the logging outside the SetupPeerVerfication call which would mean duplicating the conditional logical; adding the clientCaStack to the SetupPeerVerification signature; make logging optional and treat that as a special case where I log prior and fake the flags?

Reply:
> Passing the config flags through to underlying API seems ugly to me, but if that's what you want. OK.

Yes, that is the best _overall_ option for now AFAICT. If you prefer, you can think of it as the lesser evil: Each option has advantages and disadvantages, but the problems with all the other (posted or discussed) options are worse than problems with this option AFAICT.

> especially given it references a bunch of other SSL behaviors unrelated to peer verification

This is not a valid concern from the code architecture point of view -- the flags are currently an integer but they are essentially a configuration object and should be viewed as such for the purpose of this discussion. Each piece of code that gets a configuration object (or a relevant portion of that object) can take what it needs from it. In fact, it is easy to imagine a situation where a function like SetupPeerVerification() needs more than just peer verification flags to work correctly.

When calling SetupPeerVerification(), we could mask parsedFlags to leave just the flags currently used by SetupPeerVerification(), but I see no value in that extra effort/complication.  We could even convert the flags into an object of some new dedicated configuration class, but, again, I see no value in that extra effort/complication.

Another related important architectural aspect here is that, all other factors being equal, it is best to interpret peer validation configuration in one place. The current code splits/spreads that interpretation across two different functions. So far, we have discovered problems with that split (anonymous boolean parameters are evil and their derivation may obscure lack of error checking), but we have not discovered a valid reason to split processing in the first place.

Reply:
> How should I handle the `clientCaStack` conditional affecting the verification mode? 

Here are a few sketches. Please pick the one you dislike the least and adjust as needed.

### Lazy/minimum:

```C++
    if (clientCaStack) {
        ...
        SetupPeerVerification(ctx.get(), parsedFlags);

        updateContextCrl(ctx);
        updateContextTrust(ctx);
    } else {
        // TODO: Reject configuration flags that require a client CA
        SetupPeerVerification(ctx.get(), SSL_FLAG_NONE);
    }
```

For the second SetupPeerVerification() above, the SSL_FLAG_NONE (with a value of zero) will need to be added to the list of the already `#define`d configuration flags.

### Detects misconfigurations:

```C++
    if (clientCaStack) {
        ...
        ConfigurePeerVerification(*ctx, parsedFlags);

        updateContextCrl(ctx);
        updateContextTrust(ctx);
    } else {
        DisablePeerVerification(*ctx, parsedFlags);
    }
```

### Compact misconfiguration detection:

```C++
    if (clientCaStack) {
        ...
        updateContextCrl(ctx);
        updateContextTrust(ctx);
    }
    ConfigurePeerVerification(ctx.get(), clientCaStack, parsedFlags);
```

In this last sketch, ConfigurePeerVerification()'s second parameter is a pointer to the optional CA. The function just uses it to know whether peer validation should be allowed or denied. It can even be a `void *` pointer! I would not make it a boolean for the reasons similar to "two anonymous booleans are evil" , but it can be a boolean if really needed.

Reply:
> Yes, that is the best overall option for now AFAICT.

Enums seems the clearest and simplest to me but so be it.

> they are essentially a configuration object and should be viewed as such for the purpose of this discussion.

I understand that. I'm debating the use of app config for directly setting lower level API state. Just as I wouldn't pass ACL config into the peer verification calls I wouldn't want to pass any other non-peer verification config. It makes using it in other places more difficult / clumsy. See below.

How should I resolve the problem with clientCaStack in ::updateContextClientCa? 

If I move to using the config flags for SetupPeerVerification and move the logging as well there isn't a clean way to manage when clientCaStack = nullptr. 

1) Add clientCaStack to the SetupPeerVerification signature and do the check twice?
2) Fake flags to get SSL_VERIFY_NONE set and double log? or move the logging outside SetupPeerVerification?
3) Add a new SSL_FLAG to indicate clientCaStack = nullptr?

None of these are pretty.

Reply:
I don't see how the 'lazy' works. The behavior is that if clientCaStack is not null then verify peer is enabled unless DONT_VERIFY_PEER is set. Having SSL_FLAG_NONE be 0 is indistinguishable from the current default. It'd require swapping the DONT to DO or making NONE nonzero.

I'll just do the "detects" one.

Reply:
> [Making the 'lazy' version work would] require ... making NONE nonzero.

Agreed. SSL_FLAG_NONE must not be zero for the lazy sketch to work. Fortunately, as a new flag, it can be anything (and can be renamed to something like `SSL_FLAG_NO_CA_CONFIGURED` to avoid the customary association of NONE with zero). Sorry I missed that detail when sketching the solutions.

The introduction of a special non-zero flag makes the "lazy" option even worse than it first looked.

> I'll just do the "detects" one.

Thank you. The misconfiguration detecting sketches require more development work, but they also improve Squid ability to detect bogus configurations and configuration typos.

Reply:
Shouldn't bad config be picked up in the config parser? Not down in the usage of said config? By this point in the code there should be no misconfiguration.

Reply:
> Shouldn't bad config be picked up in the config parser?

In the ideal world: the answer would be "no": The parser should complain about syntax errors, and the errors we are dealing here are about semantics. Moreover, these errors are partially dependent on the outcomes of such complex operations as loading a certificate chain from a disk file. The latter dependencies/complexities explain why these checks are not done in the parsing code today, even though a lot of semantics checks are misplaced there.

> By this point in the code there should be no misconfiguration.

Yes, you sensed the right problem, but we are not going to solve it in this PR AFAICT. Squid configuration has several huge architectural flaws that we are slowly addressing, including the lack of clear separation between configuration parsing/validation and configuration application. When we are done, the placement of the code you are working on now will probably feel more natural, but even if it is misplaced, it would be easy to move.

You are already going the extra mile to improve the quality of this code you touch. Loading you with the herculean task of separating SSL configuration loading from SSL configuration application would be rather unfair IMO, especially since doing so will probably require changes going well beyond the code affected by the feature you are interested in.

Reply:
I should have said "config system" rather than "parser." I meant just that the config code should be doing all syntax and data validation. Hence my opposition to passing the raw config bit flags into the runtime API. If the interest is in separating these further I'm not following why furthering it is desired.

> but we are not going to solve it in this PR AFAICT. 

Nor is that what I'm trying to accomplish. In fact I already feel this PR has turned into more than it should. Fixing debug messages and whatnot should, IMO, be a separate PR and had I been asked to look into doing so out of band I'd have been happy to do so but doing it as a prerequisite for accepting a feature patch frustrates things.